### PR TITLE
Separate CD for ADOT Operator mirroring

### PR DESCRIPTION
**Pull request by erichsueh3**
**Description:** <Describe what has changed.>
This PR adds a separate CD Workflow for ADOT Operator mirroring.

**Testing:**
In Progress

### CODE REVIEW

**Comment by erichsueh3** [comment]
Couple of notes:
1. Not sure if we want this to be able to be called manually only or also be able to be called when the main CD workflow is called
2. Not sure if setting up Go 1.x is necessary
3. Will eventually also remove this workflow from the main CD workflow
4. Can anyone provide insight into how to "test" this, as all it does is mirror the ADOT Operator image to public ECR?

**Comment by bryan-aguilar** [comment]
In response to your comments
1. This should be able to be triggered by a dispatch. It may also be valid to trigger this during the Collector CD workflow also. Although I think it may be better to keep them as isolated CI/CD workflows that do not depend on each other. 
2. It is necessary to properly setup the environment before the `go run` is triggered. 
3. Agreed
4. Lets talk tomorrow. This is one of the issues that I am trying to solve at the moment.

**Comment by bryan-aguilar** [comment]
Will this be published as a github release and require a tag? See this [job](https://github.com/aws-observability/aws-otel-collector/blob/main/.github/workflows/CD.yml#L642) that we currently use. Does this need to be applied here also? If release notes are not required then we can possibly trim that job down to just [this](https://github.com/aws-observability/aws-otel-collector/blob/main/.github/workflows/CD.yml#L654) step.

**Comment by anuraaga** [comment]
We might also consider removing the existing charts there and renaming the repo to aws-otel-operator. IMO it seems a bit much to maintain charts in addition to the repo, documenting it all up to ask customers to choose makes things more complicated for them. Just having the operator as the canonical way of deploying ADOT to k8s will simplify the story a lot I think.

File: .github/workflows/CD-adot-operator.yml
Comment (jefchien) [inline .github/workflows/CD-adot-operator.yml:39]: You are no longer running the version checking script and don't have a cache, so these conditions aren't needed.

**Comment by jefchien** [comment]
@erichsueh3 You can test it in your fork by setting your own credentials as the `RELEASE_KEY_ID` and `RELEASE_SECRET`. After running it, you should see the two repos mirrored in your ECR.

If it's still going to be called by the CD and only manually run sometimes, then it makes sense to keep it in this repo. Otherwise, I agree with @anuraaga.

File: .github/workflows/CD-adot-operator.yml
Comment (jefchien) [inline .github/workflows/CD-adot-operator.yml:14]:
```suggestion
name: Mirror ADOT Operator
```

**Comment by bryan-aguilar** [comment]
> Since the operator docker images are basically an implementation detail of the helm charts, which are the actual user-facing thing, should this mirroring happen in that repo? I think putting it here makes this repo's CI even more confusing (perhaps we've passed the cliff where it doesn't matter anymore though :P)
> 
> https://github.com/aws-observability/aws-otel-helm-charts

> We might also consider removing the existing charts there and renaming the repo to aws-otel-operator. IMO it seems a bit much to maintain collector-only deployment charts in addition to the operator, documenting it all up to ask customers to choose makes things more complicated for them. Just having the operator as the canonical way of deploying ADOT to k8s will simplify the story a lot I think.

I think this is a really good idea to explore. I had briefly thought that these operator actions should exist in a repo on their own. I think the pro/cons of moving these actions should be weighed before making a final decision to merge this. I think as a first step isolating operator CI/CD into it's own repository would make sense. A future action of simplifying existing charts should be the second step.

**Comment by erichsueh3** [comment]
> Will this be published as a github release and require a tag? See this [job](https://github.com/aws-observability/aws-otel-collector/blob/main/.github/workflows/CD.yml#L642) that we currently use. Does this need to be applied here also? If release notes are not required then we can possibly trim that job down to just [this](https://github.com/aws-observability/aws-otel-collector/blob/main/.github/workflows/CD.yml#L654) step.

Don't think that we need to publish this as a release, but if anyone feels strongly the other way then we can

**Comment by erichsueh3** [comment]
> > Since the operator docker images are basically an implementation detail of the helm charts, which are the actual user-facing thing, should this mirroring happen in that repo? I think putting it here makes this repo's CI even more confusing (perhaps we've passed the cliff where it doesn't matter anymore though :P)
> > https://github.com/aws-observability/aws-otel-helm-charts
> 
> > We might also consider removing the existing charts there and renaming the repo to aws-otel-operator. IMO it seems a bit much to maintain collector-only deployment charts in addition to the operator, documenting it all up to ask customers to choose makes things more complicated for them. Just having the operator as the canonical way of deploying ADOT to k8s will simplify the story a lot I think.
> 
> I think this is a really good idea to explore. I had briefly thought that these operator actions should exist in a repo on their own. I think the pro/cons of moving these actions should be weighed before making a final decision to merge this. I think as a first step isolating operator CI/CD into it's own repository would make sense. A future action of simplifying existing charts should be the second step.

I agree with both  of you, having the Operator CI/CD in a different repository would make a lot of sense. We can document the pros and cons of this, and also document what that process might look like, and have a discussion early next week about it before making any decisions

**Comment by erichsueh3** [comment]
Update on this PR: the workflow (and the script to check if the image should be mirrored) work as intended. However, there is now the issue within the ADOT Operator mirroring tool itself; sometime in January, OTEL Operator images [stopped being mirrored](https://github.com/open-telemetry/opentelemetry-operator/pull/661/files) to quay.io, which is where the ADOT Operator is mirrored from. This means that both the release workflow and this new ad hoc workflow will not mirror images anymore, and there is now an extra step to change the tool to mirror from ghcr instead of quay which I am working on now.

**Comment by bryan-aguilar** [comment]
Hi @erichsueh3,

That's an interesting find relating to the mirroring issue. I went ahead and converted this to a draft while the work is still in progress. Feel free to toggle it back when it is ready for review.

**Comment by erichsueh3** [comment]
> Hi @erichsueh3,
> 
> That's an interesting find relating to the mirroring issue. I went ahead and converted this to a draft while the work is still in progress. Feel free to toggle it back when it is ready for review.

Hey @bryan-aguilar, sounds good. Will work on a separate PR for the mirroring tools

File: .github/workflows/CD-adot-operator.yml
Comment (bryan-aguilar) [inline .github/workflows/CD-adot-operator.yml:32]:
```suggestion
          go-version: ^1.17.7
```

File: .github/workflows/CD-adot-operator.yml
Comment (bryan-aguilar) [inline .github/workflows/CD-adot-operator.yml:43]: Is this collector or operator version? Does this need to go into the CD workflow?

Comment (erichsueh3) [reply .github/workflows/CD-adot-operator.yml]: This is the Operator version, and I don't think it needs to go in the CD workflow since it should only run on release and also checks collector version; also this should be `tools/release/adot-operator-images-mirror/version-check.sh`, fixed that just now.

File: .github/workflows/CD-adot-operator.yml
Comment (erichsueh3) [reply .github/workflows/CD-adot-operator.yml]: Fixed

File: .github/workflows/CD.yml
Comment (bryan-aguilar) [inline .github/workflows/CD.yml:641]:
```suggestion
        if: ${{ github.event.inputs.skip-adot-operator-mirroring == 'false' && (steps.release-latest-image.outputs.cache-hit != 'true' && (steps.version.outputs.patch-update == 'false') }}
```

Can you verify that this logic is the same by referencing [this](https://github.com/aws-observability/aws-otel-collector/blob/main/tools/workflow/docker-version-compare.sh) script.

Comment (erichsueh3) [reply .github/workflows/CD.yml:641]: Not sure why it checks for `steps.version.outputs.same-version == 'true'` so that should be removed; checking only `steps.version.outputs.patch-update == 'false'` doesn't account for if major/minor updates have happened. If we check it as `steps.version.outputs.patch-update == 'true'` though it should be equivalent, while also adding a check for patch updates (if we want that in adot operator images)

Comment (bryan-aguilar) [reply .github/workflows/CD.yml:641]: I believe only `xxx-update` values can be true at once. I believe `any-update` can also be true but that is not referenced here.

**Comment by bryan-aguilar** [comment]
The plan is to merge this PR now and investigate moving the ADOT operator releases to a separate repository in the futures. Merging this now will unblock the ADOT Operator being out of sync with upstream.

**Review by bryan-aguilar — APPROVED** [review]

**Pull request merged**
